keps/sig-api-machinery/20191210-consistent-reads-from-cache.md

```diff
@@ -0,0 +1,294 @@
+We will benchmark and select the most efficient of these options.
+
+#### Alternative 2: Use WatchProgressRequest to request watch updates when needed
+
+etcd 3.4+ provides a `WatchProgressRequest` request that can be made on a watch channel. When requested,
+etcd will send a progress event on that watch as soon as possible.
+
+When an consistent read request is received and the watch cache is enabled:
+- Get the current revision from etcd using a range read with limit=0, just like in alternative 1.
+- If the cache already has the current revision, serve the request from cache
+- If the current revision is not in the cache:
+  - Send a `WatchProgressRequest` to etcd on the watch channel that the watch cache is consuming.
+- Use the existing waitUntilFreshAndBlock function in the watch cache to wait for the watch to catch up to the current revision
+- If the block times out, skip the cache and serve the request directly from storage.
+
+This alternative requires using `WatchProgressRequest` which is only available in etcd 3.4+, and so would
+require we make the kube-apiserver aware of etcd's minor version, which is described in more detail later.
+
+#### Comparing the alternatives
+
+We will experiment with modifying etcd to automatically send progress events in
+short intervals of 20ms or so to see if this keeps the latest observed revision
+of watch caches sufficiently up-to-date that they rarely need to fall back to
+serving resourceVersion="" requests directly from etcd.  We will check for cache
+hit ratios and check scalability. If these are promising we will then explore
+what the optimal interval to have etcd send progress events is. If this works
+well enough we will go with alternative 1 and will not need to implement
+alternative 2.
+
+We will also explore how long the kube-apiserver should wait for a watch cache
+to catch up to the needed revision. I.e. for a progress event interval of 20ms,
+should the kube-apiserver also wait 20ms to a desired revision to become
+available or should it fallback to making the request to etcd sooner?
+
+Optional: For some (but not all) of the etcd progress watch events, also create a
+kubernetes "bookmark" watch event and send it to kube-apiserver clients so that
+reflectors and shared informers are kept up-to-date. The benefit of this is that
+it minimizes the chance that these clients will end up with an out-of-date
+resource version and need to relist (which can impact scalability). See [Watch
+Bookmarks](https://github.com/kubernetes/enhancements/blob/master/keps/sig-api-machinery/20190206-watch-bookmark.md)
+for details.
+
+### Make the kube-apiserver aware of etcd's minor version
+
+This is only needed if we go with Alternative 2 (Use WatchProgressRequest to
+request watch updates when needed) since Alterative 1 is compatible with all
+etcd 3.x versions.
+
+Allow the etcd minor version to be specified in the kube-apiserver
+`--storage-backend` flag, e.g. `--storage-backend=etcd3.4`.  When client
```

Comment:
I'd really prefer not to do this if at all possible.

Reply:
Agreed. It looks like we're starting to strongly favor alternative 1, where we don't need to send etcd `WatchProgressRequest`s at all. I'll remove alternative 2 from this KEP and remove this section as well.